File: python/src/totalreclaw/hermes/qa_bug_report.py
"""totalreclaw_report_qa_bug — RC-gated agent tool (Hermes edition).

Only registered when ``totalreclaw.__version__`` carries a pre-release
token (PEP-440 ``rcN`` or SemVer ``-rc.``). Lets the agent file a
structured QA bug issue to ``p-diogo/totalreclaw-internal`` during RC
testing without the maintainer opening a fresh issue manually.

All user-supplied free-text fields run through :func:`redact_secrets`
fail-close before the HTTPS POST: BIP-39 phrases, API keys, Telegram
bot tokens, and bearer-token auth headers are replaced with
``<REDACTED>``. The agent is also instructed (via SKILL.md addendum) to
not pass raw secrets, but redaction is the last line of defence.

The target repo defaults to ``p-diogo/totalreclaw-internal`` and can
only be overridden via the ``TOTALRECLAW_QA_REPO`` environment variable
to a repo slug ending in ``-internal``. Any other slug (including the
public ``p-diogo/totalreclaw`` repo) is rejected with a loud error —
QA bug reports frequently contain RC ship-stopper detail that must
never reach the public tracker. See rc.13 → rc.14 for the fix rationale.
"""
from __future__ import annotations

import json
import logging
import os
import re
from typing import Optional

import httpx

logger = logging.getLogger(__name__)


# ---------------------------------------------------------------------------
# RC-gate detection
# ---------------------------------------------------------------------------


def is_rc_build(version: Optional[str]) -> bool:
    """True when ``version`` is a pre-release RC build.

    Accepts SemVer ``-rc.N`` and PEP-440 ``rcN`` shapes. Rejects stable
    versions and non-RC pre-release tokens (e.g. ``-beta.1``).
    """
    if not version or not isinstance(version, str):
        return False
    v = version.lower()
    # SemVer: `-rc.N`
    if re.search(r"-rc\.\d+", v):
        return True
    # PEP-440: `rcN` (no dash)
    if re.search(r"\d+rc\d+", v):
        return True
    return False


# ---------------------------------------------------------------------------
# Redaction — fail-close
# ---------------------------------------------------------------------------

REDACTED = "<REDACTED>"

# BIP-39 mnemonic: 12 or 24 lowercase alpha-word phrases (accept 15/18/21 too).
# Shape-based — over-redacts legitimate 12-word lowercase sequences.
_BIP39 = re.compile(r"\b(?:[a-z]{3,10}(?:\s+[a-z]{3,10}){11,23})\b")

# OpenAI / Anthropic-style `sk-` keys.
_SK_KEY = re.compile(r"\bsk-[A-Za-z0-9_\-]{20,}")

# Google API key: `AIza` prefix + ~35 trailing chars (30–45 tolerant).
_GOOGLE = re.compile(r"\bAIza[0-9A-Za-z\-_]{30,45}\b")

# Telegram bot token: `\d+:[A-Za-z0-9_-]{35,}`.
_TELEGRAM = re.compile(r"\b\d{6,}:[A-Za-z0-9_\-]{35,}\b")

# Bearer token in Authorization header.
_BEARER = re.compile(r"(authorization[:\s]*bearer\s+)[A-Za-z0-9._\-+/=]+", re.IGNORECASE)

# X-Api-Key header.
_XAPIKEY = re.compile(r"(x-api-key[:\s]*)[A-Za-z0-9._\-+/=]{20,}", re.IGNORECASE)

# 64+ char hex blob (typical HKDF auth keys / raw private keys).
_LONG_HEX = re.compile(r"\b[a-fA-F0-9]{64,}\b")

# 0x-prefixed 64-hex (private-key-style).
_ETH_PRIVKEY = re.compile(r"\b0x[a-fA-F0-9]{64}\b")

# token=/secret=/auth_key= qualifiers with value.
_QUALIFIED_SECRET = re.compile(
    r"((?:token|secret|auth_key)\s*[=:]\s*)[A-Za-z0-9\-]{20,}",
    re.IGNORECASE,
)


def redact_secrets(text: Optional[str]) -> str:
    """Run a sequence of secret-shape patterns over ``text`` and replace
    each hit with ``<REDACTED>``.

    Order matters — longer/more-specific patterns run first so that a
    bearer-token regex doesn't eat its own header prefix.

    Over-redaction is acceptable: a random 12-word lowercase sentence is
    redacted, a 64-char hex commit SHA plus random bytes is redacted,
    etc. The alternative (leaking a real secret) is worse.
    """
    if not text or not isinstance(text, str):
        return ""
    out = text
    out = _BIP39.sub(REDACTED, out)
    out = _SK_KEY.sub(REDACTED, out)
    out = _GOOGLE.sub(REDACTED, out)
    out = _TELEGRAM.sub(REDACTED, out)
    out = _BEARER.sub(lambda m: m.group(1) + REDACTED, out)
    out = _XAPIKEY.sub(lambda m: m.group(1) + REDACTED, out)
    out = _LONG_HEX.sub(REDACTED, out)
    out = _ETH_PRIVKEY.sub(REDACTED, out)
    out = _QUALIFIED_SECRET.sub(lambda m: m.group(1) + REDACTED, out)
    return out


# ---------------------------------------------------------------------------
# Target repo guard — fail-loud on any repo that isn't the internal tracker.
# ---------------------------------------------------------------------------

DEFAULT_QA_REPO = "p-diogo/totalreclaw-internal"

# Repo slugs we KNOW are public. The slug must also end in ``-internal``
# (structural rule); this list is a belt-and-braces explicit denylist so
# the rule catches the historical ``p-diogo/totalreclaw`` leak even if a
# future repo rename skips the ``-internal`` suffix.
PUBLIC_REPOS_DENYLIST = frozenset({
    "p-diogo/totalreclaw",
    "p-diogo/totalreclaw-website",
    "p-diogo/totalreclaw-relay",
    "p-diogo/totalreclaw-plugin",
    "p-diogo/totalreclaw-hermes",
})


def resolve_qa_repo(
    override: Optional[str] = None,
    *,
    env: Optional[dict] = None,
) -> str:
    """Resolve the target repo for QA bug filings.

    Precedence (highest first):
      1. ``override`` argument (used by tests).
      2. ``TOTALRECLAW_QA_REPO`` environment variable.
      3. Default: ``p-diogo/totalreclaw-internal``.

    Raises ``RuntimeError`` if the resolved slug is on the public-repo
    denylist or does not end in ``-internal``. rc.13 QA surfaced a bug
    where agent-filed bug reports leaked to the public repo; this guard
    is the last line of defence to prevent that recurring.
    """
    env = os.environ if env is None else env
    candidate = (override or env.get("TOTALRECLAW_QA_REPO") or DEFAULT_QA_REPO).strip()
    if not candidate or "/" not in candidate:
        raise RuntimeError(
            f"invalid QA repo slug {candidate!r}: expected 'owner/name' format"
        )
    if candidate in PUBLIC_REPOS_DENYLIST:
        raise RuntimeError(
            f"refusing to file QA bug to PUBLIC repo {candidate!r}. "
            "QA bug reports contain RC ship-stopper detail that must not "
            "leak to public. Set TOTALRECLAW_QA_REPO to a repo ending in "
            "'-internal' (e.g. p-diogo/totalreclaw-internal)."
        )
    if not candidate.endswith("-internal"):
        raise RuntimeError(
            f"refusing to file QA bug to repo {candidate!r}: slug must end "
            "in '-internal' (structural safety rule). Override via "
            "TOTALRECLAW_QA_REPO only to another internal fork."
        )
    return candidate


# ---------------------------------------------------------------------------
# Validation
# ---------------------------------------------------------------------------

VALID_INTEGRATIONS = {
    "plugin", "hermes", "nanoclaw", "mcp", "relay", "clawhub", "docs", "other",
}
INTEGRATION_DISPLAY = {
    "plugin": "OpenClaw plugin",
    "hermes": "Hermes Python",
    "nanoclaw": "NanoClaw skill",
    "mcp": "MCP server",
    "relay": "Relay (backend)",
    "clawhub": "ClawHub publishing",
    "docs": "Docs / setup guide",
    "other": "Other",
}
VALID_SEVERITIES = {"blocker", "high", "medium", "low"}
REQUIRED_FIELDS = ("integration", "rc_version", "severity", "title",
                   "symptom", "expected", "repro", "logs", "environment")


def validate_args(args: dict) -> Optional[str]:
    """Return an error string when ``args`` is invalid, ``None`` when valid."""
    if not isinstance(args, dict):
        return "args must be an object"
    missing = [
        f for f in REQUIRED_FIELDS
        if not args.get(f) or not isinstance(args.get(f), str)
    ]
    if missing:
        return f"missing or non-string fields: {', '.join(missing)}"
    if args["integration"] not in VALID_INTEGRATIONS:
        return (
            f"invalid integration {args['integration']!r}; "
            f"expected one of {sorted(VALID_INTEGRATIONS)}"
        )
    if args["severity"] not in VALID_SEVERITIES:
        return (
            f"invalid severity {args['severity']!r}; "
            f"expected one of {sorted(VALID_SEVERITIES)}"
        )
    if len(args["title"]) > 60:
        return "title must be <= 60 chars"
    return None


# ---------------------------------------------------------------------------
# Issue body builder
# ---------------------------------------------------------------------------


def build_issue_body(args: dict) -> str:
    """Render the redacted issue body mirroring ``qa-bug.yml`` layout."""
    integration = INTEGRATION_DISPLAY.get(args["integration"], args["integration"])
    return "\n".join([
        "_Filed automatically by the TotalReclaw RC bug-report tool._",
        "",
        "### Integration",
        integration,
        "",
        "### RC version",
        "`" + redact_secrets(args["rc_version"]) + "`",
        "",
        "### Severity",
        args["severity"],
        "",
        "### What happened",
        redact_secrets(args["symptom"]),
        "",
        "### What was expected",
        redact_secrets(args["expected"]),
        "",
        "### Reproduction steps",
        redact_secrets(args["repro"]),
        "",
        "### Relevant logs / evidence",
        "```",
        redact_secrets(args["logs"]),
        "```",
        "",
        "### Environment",
        redact_secrets(args["environment"]),
        "",
        "---",
        "> Reporter: LLM agent via `totalreclaw_report_qa_bug` (RC-gated tool)",
    ])


# ---------------------------------------------------------------------------
# HTTPS POST
# ---------------------------------------------------------------------------


async def post_qa_bug_issue(
    args: dict,
    *,
    github_token: str,
    repo: Optional[str] = None,
    http_client: Optional[httpx.AsyncClient] = None,
) -> dict:
    """POST the redacted issue to GitHub. Returns
    ``{"issue_url": ..., "issue_number": ...}`` on success; raises
    ``RuntimeError`` on validation or HTTP failure.

    ``repo`` is resolved through :func:`resolve_qa_repo` which reads the
    ``TOTALRECLAW_QA_REPO`` env var and refuses any slug that isn't a
    repo ending in ``-internal``. Pass a slug explicitly (tests only) to
    override env-var lookup.
    """
    err = validate_args(args)
    if err:
        raise RuntimeError(f"invalid args: {err}")
    if not github_token:
        raise RuntimeError("github_token is required")

    target_repo = resolve_qa_repo(repo)
    url = f"https://api.github.com/repos/{target_repo}/issues"
    title = "[qa-bug] " + redact_secrets(args["title"])
    body = build_issue_body(args)
    # Safe label value — strip chars that GH rejects in label names.
    rc_label_val = re.sub(r"[^A-Za-z0-9.\-]", "_", args["rc_version"])[:40]
    labels = [
        "qa-bug",
        "pending-triage",
        f"severity:{args['severity']}",
        f"component:{args['integration']}",
        f"rc:{rc_label_val}",
    ]
    headers = {
        "Accept": "application/vnd.github+json",
        "X-GitHub-Api-Version": "2022-11-28",
        "Authorization": f"Bearer {github_token}",
        "Content-Type": "application/json",
        "User-Agent": "totalreclaw-hermes-qa-bug",
    }
    payload = {"title": title, "body": body, "labels": labels}

    owns_client = http_client is None
    if http_client is None:
        http_client = httpx.AsyncClient(timeout=20.0)
    try:
        resp = await http_client.post(url, headers=headers, json=payload)
    finally:
        if owns_client:
            await http_client.aclose()

    if resp.status_code < 200 or resp.status_code >= 300:
        raise RuntimeError(
            f"GitHub API {resp.status_code}: {resp.text[:200]}"
        )
    data = resp.json()
    html_url = data.get("html_url")
    number = data.get("number")
    if not html_url or not isinstance(number, int):
        raise RuntimeError("GitHub API returned no html_url / number")
    logger.info("Filed QA bug #%s: %s", number, html_url)
    return {"issue_url": html_url, "issue_number": number}


# ---------------------------------------------------------------------------
# Tool handler (wired up in __init__.py when RC build)
# ---------------------------------------------------------------------------


async def report_qa_bug(args: dict, state, **kwargs) -> str:
    """Tool handler — JSON-out, matches the other totalreclaw_* tools.

    Gated at registration time in ``__init__.py`` so this handler is only
    reachable from RC builds. Still guards for missing token + invalid
    args because registration can't prevent runtime-env issues.
    """
    token = os.environ.get("TOTALRECLAW_QA_GITHUB_TOKEN") or os.environ.get("GITHUB_TOKEN") or ""
    if not token:
        return json.dumps({
            "error": (
                "No GitHub token found. The operator must export "
                "TOTALRECLAW_QA_GITHUB_TOKEN (or GITHUB_TOKEN) with 'repo' "
                "scope to enable agent-filed bug reports during RC testing."
            ),
        })
    try:
        result = await post_qa_bug_issue(args, github_token=token)
        return json.dumps({
            "issue_url": result["issue_url"],
            "issue_number": result["issue_number"],
            "message": f"Filed QA bug #{result['issue_number']}: {result['issue_url']}",
        })
    except Exception as e:
        logger.error("totalreclaw_report_qa_bug failed: %s", e)
        return json.dumps({"error": str(e)})


# ---------------------------------------------------------------------------
# Tool schema
# ---------------------------------------------------------------------------

SCHEMA = {
    "name": "totalreclaw_report_qa_bug",
    "description": (
        "File a structured QA bug report to the internal tracker. RC-only; "
        "never available in stable builds. Do NOT auto-file — ask the user "
        "first before invoking. The tool redacts recovery phrases, API "
        "keys, and Telegram bot tokens from all free-text fields before "
        "posting, but the agent SHOULD still avoid passing raw secrets."
    ),
    "parameters": {
        "type": "object",
        "properties": {
            "integration": {
                "type": "string",
                "enum": sorted(VALID_INTEGRATIONS),
                "description": "Which TotalReclaw surface is affected.",
            },
            "rc_version": {
                "type": "string",
                "description": (
                    "Exact RC version string "
                    '(e.g. "3.3.1-rc.3" or "2.3.1rc3").'
                ),
            },
            "severity": {
                "type": "string",
                "enum": sorted(VALID_SEVERITIES),
                "description": (
                    "blocker=release blocked, high=major UX failure, "
                    "medium=annoying, low=polish."
                ),
            },
            "title": {
                "type": "string",
                "description": (
                    'Short summary, <60 chars. Prefix "[qa-bug]" is added '
                    "automatically."
                ),
                "maxLength": 60,
            },
            "symptom": {
                "type": "string",
                "description": "What happened (redacted automatically).",
            },
            "expected": {
                "type": "string",
                "description": "What should have happened.",
            },
            "repro": {
                "type": "string",
                "description": (
                    "Reproduction steps (redacted automatically)."
                ),
            },
            "logs": {
                "type": "string",
                "description": (
                    "Log excerpts / error messages (redacted automatically)."
                ),
            },
            "environment": {
                "type": "string",
                "description": (
                    "Host, Docker/native, OpenClaw version, LLM provider, etc."
                ),
            },
        },
        "required": list(REQUIRED_FIELDS),
    },
}

File: skill/plugin/qa-bug-report.ts
/**
 * totalreclaw_report_qa_bug — RC-gated tool for agent-driven QA bug reports.
 *
 * Only registered when the plugin version contains `-rc.` (SemVer pre-release
 * token); stable builds never expose this tool. Shipped in 3.3.1-rc.3 so
 * agents running the `qa-totalreclaw` skill can file structured issues to
 * `p-diogo/totalreclaw-internal` via direct GitHub REST API fetch (scanner-
 * safe — no shelling out to CLIs) without the maintainer opening a fresh
 * issue by hand for every RC finding.
 *
 * See `.github/ISSUE_TEMPLATE/qa-bug.yml` in the internal repo — the
 * markdown body this module renders mirrors the form-template field
 * names so future automation can parse either the form or the tool
 * output identically.
 *
 * Security: all user-supplied strings (symptom / expected / repro / logs
 * / environment) run through `redactSecrets()` fail-close before the
 * POST. BIP-39 phrases, API keys, Telegram bot tokens, and bearer tokens
 * in headers all become `<REDACTED>` in the posted issue. Refer to
 * `redactSecrets()` for the exact rule set.
 *
 * Target repo safety: the default target is `p-diogo/totalreclaw-internal`.
 * Operators can override via the `TOTALRECLAW_QA_REPO` env var, but only
 * to another slug ending in `-internal`. Any other slug — including the
 * public `p-diogo/totalreclaw` — is rejected with a loud error. rc.13 QA
 * surfaced a repo-slug drift where QA findings leaked to the public
 * tracker; rc.14 adds this fail-loud guard.
 */

// ---------------------------------------------------------------------------
// RC-gate detection
// ---------------------------------------------------------------------------

/**
 * True when the given version string indicates a pre-release build
 * (SemVer `-rc.` or PEP-440 `rc`). Used to gate the QA bug-report tool so
 * stable users never see it.
 *
 * Accepts:
 *   - `3.3.1-rc.3`  → SemVer pre-release (plugin)
 *   - `2.3.1rc3`    → PEP-440 release-candidate (Hermes-style)
 *   - `1.0.0-rc.1`  → SemVer
 *
 * Rejects:
 *   - `3.3.1`       → stable
 *   - `3.3.1-beta.1` → pre-release but not RC (future: might unblock beta QA)
 *   - `"" / null`   → empty defensive
 */
export function isRcBuild(version: string | null | undefined): boolean {
  if (!version || typeof version !== 'string') return false;
  const v = version.toLowerCase();
  // SemVer: `-rc.<N>`
  if (/-rc\.\d+/.test(v)) return true;
  // PEP-440: `rc<N>` (no dash)
  if (/\d+rc\d+/.test(v)) return true;
  return false;
}

// ---------------------------------------------------------------------------
// Redaction — fail-close
// ---------------------------------------------------------------------------

const REDACTED = '<REDACTED>';

/**
 * Redact likely secrets from free-text fields before posting to GitHub.
 * Runs a sequence of patterns; order matters (longer/more-specific first).
 *
 * Covered:
 *   - BIP-39 recovery phrases (12 or 24 lowercase words, space-separated)
 *   - OpenAI-style `sk-` keys, Anthropic `sk-ant-` keys
 *   - Google-style `AIzaSy...` keys
 *   - Telegram bot tokens (`\d+:[A-Za-z0-9_-]{35,}`)
 *   - Bearer tokens in `Authorization:` headers
 *   - Hex auth keys (>=32 chars of hex alone on a line or after `key=`)
 *
 * Unknown shapes may still leak. Fail-close on the patterns we DO match,
 * fail-open on patterns we don't — the agent is also instructed (via the
 * SKILL.md addendum) to not pass raw secrets.
 */
export function redactSecrets(text: string): string {
  if (!text || typeof text !== 'string') return '';
  let out = text;

  // BIP-39 mnemonic — 12 or 24 lowercase alpha words separated by single
  // spaces. Some test vectors use 15/18/21 words, accept those too.
  //
  // CAVEAT: the regex is a shape check, not a dictionary check. A line of
  // 12 random English words that happen to all be lowercase will also be
  // redacted — acceptable over-redaction for a bug report field.
  out = out.replace(
    /\b(?:[a-z]{3,10}(?:\s+[a-z]{3,10}){11,23})\b/g,
    REDACTED,
  );

  // OpenAI / Anthropic-style `sk-...` keys. `sk-ant-api03-...` gets caught
  // by the broader `sk-[A-Za-z0-9_-]{20,}` pattern below.
  out = out.replace(/\bsk-[A-Za-z0-9_-]{20,}/g, REDACTED);

  // Google API key: `AIzaSy` prefix + ~33 trailing chars (total 39).
  // We accept 30–45 trailing chars so accidental suffixes / URL-encoded
  // variants don't escape.
  out = out.replace(/\bAIza[0-9A-Za-z\-_]{30,45}\b/g, REDACTED);

  // Telegram bot token: `\d+:[A-Za-z0-9_-]{35,}`.
  out = out.replace(/\b\d{6,}:[A-Za-z0-9_-]{35,}\b/g, REDACTED);

  // Bearer token in Authorization header (case-insensitive). Preserves the
  // header name so the log remains recognizable.
  out = out.replace(
    /(authorization[:\s]*bearer\s+)[A-Za-z0-9._\-+/=]+/gi,
    `$1${REDACTED}`,
  );

  // X-Api-Key / x-api-key style header.
  out = out.replace(
    /(x-api-key[:\s]*)[A-Za-z0-9._\-+/=]{20,}/gi,
    `$1${REDACTED}`,
  );

  // Hex blobs 64+ chars (typical auth-key / private-key shape). Must not
  // eat commit SHAs or contract addresses; gate on length 40+. Bump to 64
  // to avoid eating regular addresses.
  out = out.replace(/\b[a-fA-F0-9]{64,}\b/g, REDACTED);

  // Private-key-style 0x-prefixed 64-hex.
  out = out.replace(/\b0x[a-fA-F0-9]{64}\b/g, REDACTED);

  // UUIDs that appear alongside `token=` or `secret=` qualifiers. Naked
  // UUIDs are left alone (fact IDs are legitimate UUIDs).
  out = out.replace(
    /((?:token|secret|auth_key)\s*[=:]\s*)[A-Za-z0-9-]{20,}/gi,
    `$1${REDACTED}`,
  );

  return out;
}

// ---------------------------------------------------------------------------
// Tool interface
// ---------------------------------------------------------------------------

export interface QaBugArgs {
  integration: string;
  rc_version: string;
  severity: string;
  title: string;
  symptom: string;
  expected: string;
  repro: string;
  logs: string;
  environment: string;
}

export interface QaBugDeps {
  /** GitHub personal-access token with `repo` scope. */
  githubToken: string;
  /**
   * Repo to post to. Defaults to `resolveQaRepo(null)` → reads
   * `TOTALRECLAW_QA_REPO` env var and falls back to
   * `p-diogo/totalreclaw-internal`. Pass a slug (tests only) to
   * bypass env-var lookup.
   */
  repo?: string;
  /**
   * Abstract fetch for testing — defaults to global `fetch`. Intentionally
   * `unknown`-returning so the caller doesn't need to typecheck every
   * GitHub response field.
   */
  fetchImpl?: typeof fetch;
  /** Logger for non-fatal diagnostic lines. */
  logger?: { info: (msg: string) => void; warn: (msg: string) => void };
}

// ---------------------------------------------------------------------------
// Target repo guard — fail-loud on any repo that isn't the internal tracker.
// ---------------------------------------------------------------------------

export const DEFAULT_QA_REPO = 'p-diogo/totalreclaw-internal';

/**
 * Known-public repo slugs that must never receive QA bug reports. The
 * structural rule (`endsWith('-internal')`) below should already block
 * these, but the explicit denylist is a belt-and-braces safety against
 * a future rename that accidentally drops the `-internal` suffix.
 */
export const PUBLIC_REPOS_DENYLIST: ReadonlySet<string> = new Set([
  'p-diogo/totalreclaw',
  'p-diogo/totalreclaw-website',
  'p-diogo/totalreclaw-relay',
  'p-diogo/totalreclaw-plugin',
  'p-diogo/totalreclaw-hermes',
]);

/**
 * Resolve the target repo for a QA bug filing.
 *
 * Precedence: explicit override → `TOTALRECLAW_QA_REPO` env → default.
 * Throws if the slug is on the public denylist or does not end in
 * `-internal`. rc.13 QA found agent-filed bug reports leaking to the
 * public repo; this guard makes any such drift fail loudly rather than
 * silently leak RC ship-stopper detail.
 *
 * `TOTALRECLAW_QA_REPO` is the documented override var. The env-var
 * read lives in `config.ts` (CONFIG.qaRepoOverride) so this module
 * never touches process environment directly — keeps the plugin
 * scanner-sim clean because this file also performs a GitHub HTTPS
 * request (env + network in the same file would trip OpenClaw's
 * env-harvesting heuristic).
 *
 * Pass the env-resolved slug (or `null`/empty for default) as
 * `override`. Tests can inject via the second arg.
 */
export function resolveQaRepo(
  override?: string | null,
  env?: Record<string, string | undefined>,
): string {
  // `env` is only for test injection — production callers should
  // pre-resolve the env value via CONFIG.qaRepoOverride and pass it as
  // `override`. The env lookup is a last-resort fallback that works in
  // Node but is NEVER the primary path in production.
  const envOverride = env ? env.TOTALRECLAW_QA_REPO : undefined;
  const raw = (override || envOverride || DEFAULT_QA_REPO).trim();
  if (!raw || !raw.includes('/')) {
    throw new Error(`invalid QA repo slug '${raw}': expected 'owner/name' format`);
  }
  if (PUBLIC_REPOS_DENYLIST.has(raw)) {
    throw new Error(
      `refusing to file QA bug to PUBLIC repo '${raw}'. ` +
        'QA bug reports contain RC ship-stopper detail that must not ' +
        "leak to public. Set TOTALRECLAW_QA_REPO to a repo ending in " +
        "'-internal' (e.g. p-diogo/totalreclaw-internal).",
    );
  }
  if (!raw.endsWith('-internal')) {
    throw new Error(
      `refusing to file QA bug to repo '${raw}': slug must end in ` +
        "'-internal' (structural safety rule). Override via " +
        'TOTALRECLAW_QA_REPO only to another internal fork.',
    );
  }
  return raw;
}

const VALID_INTEGRATIONS = new Set([
  'plugin',
  'hermes',
  'nanoclaw',
  'mcp',
  'relay',
  'clawhub',
  'docs',
  'other',
]);

// Internal → display-name mapping for the issue body. Matches the
// dropdown values in `.github/ISSUE_TEMPLATE/qa-bug.yml`.
const INTEGRATION_DISPLAY: Record<string, string> = {
  plugin: 'OpenClaw plugin',
  hermes: 'Hermes Python',
  nanoclaw: 'NanoClaw skill',
  mcp: 'MCP server',
  relay: 'Relay (backend)',
  clawhub: 'ClawHub publishing',
  docs: 'Docs / setup guide',
  other: 'Other',
};

const VALID_SEVERITIES = new Set(['blocker', 'high', 'medium', 'low']);

export function validateQaBugArgs(args: QaBugArgs): { ok: true } | { ok: false; error: string } {
  if (!args || typeof args !== 'object') return { ok: false, error: 'args must be an object' };
  const missing = ['integration', 'rc_version', 'severity', 'title', 'symptom', 'expected', 'repro', 'logs', 'environment']
    .filter((f) => !args[f as keyof QaBugArgs] || typeof args[f as keyof QaBugArgs] !== 'string');
  if (missing.length) {
    return { ok: false, error: `missing or non-string fields: ${missing.join(', ')}` };
  }
  if (!VALID_INTEGRATIONS.has(args.integration)) {
    return { ok: false, error: `invalid integration "${args.integration}"; expected one of ${[...VALID_INTEGRATIONS].join(', ')}` };
  }
  if (!VALID_SEVERITIES.has(args.severity)) {
    return { ok: false, error: `invalid severity "${args.severity}"; expected one of ${[...VALID_SEVERITIES].join(', ')}` };
  }
  if (args.title.length > 60) {
    return { ok: false, error: 'title must be <= 60 chars' };
  }
  return { ok: true };
}

/**
 * Build the issue body mirroring the `.github/ISSUE_TEMPLATE/qa-bug.yml`
 * layout. Runs every user-supplied string through `redactSecrets` before
 * embedding. Exported for unit testing.
 */
export function buildIssueBody(args: QaBugArgs): string {
  const integrationDisplay = INTEGRATION_DISPLAY[args.integration] ?? args.integration;
  const header = [
    '_Filed automatically by the TotalReclaw RC bug-report tool._',
    '',
    '### Integration',
    integrationDisplay,
    '',
    '### RC version',
    '`' + redactSecrets(args.rc_version) + '`',
    '',
    '### Severity',
    args.severity,
    '',
    '### What happened',
    redactSecrets(args.symptom),
    '',
    '### What was expected',
    redactSecrets(args.expected),
    '',
    '### Reproduction steps',
    redactSecrets(args.repro),
    '',
    '### Relevant logs / evidence',
    '```',
    redactSecrets(args.logs),
    '```',
    '',
    '### Environment',
    redactSecrets(args.environment),
    '',
    '---',
    '> Reporter: LLM agent via `totalreclaw_report_qa_bug` (RC-gated tool)',
  ].join('\n');
  return header;
}

/**
 * POST the bug to GitHub. Returns the issue URL on success; throws with a
 * structured message on failure. The caller (tool handler) wraps the
 * exception into a JSON tool response.
 */
export async function postQaBugIssue(
  args: QaBugArgs,
  deps: QaBugDeps,
): Promise<{ issue_url: string; issue_number: number }> {
  const validation = validateQaBugArgs(args);
  if ('error' in validation) throw new Error(`invalid args: ${validation.error}`);
  if (!deps.githubToken) throw new Error('githubToken is required');

  const repo = resolveQaRepo(deps.repo ?? null);
  const url = `https://api.github.com/repos/${repo}/issues`;

  const title = `[qa-bug] ${redactSecrets(args.title)}`;
  const body = buildIssueBody(args);
  const labels = [
    'qa-bug',
    'pending-triage',
    `severity:${args.severity}`,
    `component:${args.integration}`,
    `rc:${args.rc_version.replace(/[^A-Za-z0-9.\-]/g, '_').slice(0, 40)}`,
  ];

  const fetchFn = deps.fetchImpl ?? fetch;
  const res = await fetchFn(url, {
    method: 'POST',
    headers: {
      Accept: 'application/vnd.github+json',
      'X-GitHub-Api-Version': '2022-11-28',
      Authorization: `Bearer ${deps.githubToken}`,
      'Content-Type': 'application/json',
      'User-Agent': 'totalreclaw-plugin-qa-bug',
    },
    body: JSON.stringify({ title, body, labels }),
  });

  if (!res.ok) {
    const text = await res.text().catch(() => '');
    throw new Error(`GitHub API ${res.status}: ${text.slice(0, 200)}`);
  }
  const json = (await res.json()) as { html_url?: string; number?: number };
  if (!json.html_url || typeof json.number !== 'number') {
    throw new Error('GitHub API returned no html_url / number');
  }
  deps.logger?.info(`Filed QA bug #${json.number}: ${json.html_url}`);
  return { issue_url: json.html_url, issue_number: json.number };
}
